redbot/cogs/mutes/mutes.py

```diff
@@ -0,0 +1,782 @@
+                        member,
+                        author,
+                        _("Automatic unmute"),
+                        until=None,
+                    )
+                except RuntimeError as e:
+                    log.error(_("Error creating modlog case"), exc_info=e)
+        except discord.errors.Forbidden:
+            return
+
+    async def _handle_channel_unmutes(self):
+        log.debug("Checking channel unmutes")
+        for c_id, mutes in self._channel_mutes.items():
+            to_remove = []
+            channel = self.bot.get_channel(c_id)
+            if channel is None:
+                continue
+            for u_id, data in mutes.items():
+                time_to_unmute = data["until"] - datetime.utcnow().timestamp()
+                if time_to_unmute < 120.0:
+                    self._unmute_tasks[f"{c_id}{u_id}"] = asyncio.create_task(
+                        self._auto_channel_unmute_user(channel, data)
+                    )
+            for u_id in to_remove:
+                del self._channel_mutes[c_id][u_id]
+            await self.config.channel(channel).muted_users.set(self._channel_mutes[c_id])
+
+    async def _auto_channel_unmute_user(self, channel: discord.TextChannel, data: dict):
+        delay = 120 - (data["until"] - datetime.utcnow().timestamp())
+        if delay < 1:
+            delay = 0
+        await asyncio.sleep(delay)
+        try:
+            member = channel.guild.get_member(data["member"])
+            author = channel.guild.get_member(data["author"])
+            if not member or not author:
+                return
+            success, message = await self.channel_unmute_user(
+                channel.guild, channel, author, member, _("Automatic unmute")
+            )
+            if success:
+                try:
+                    await modlog.create_case(
+                        self.bot,
+                        channel.guild,
+                        datetime.utcnow(),
+                        "cunmute",
+                        member,
+                        author,
+                        _("Automatic unmute"),
+                        until=None,
+                    )
+                except RuntimeError as e:
+                    log.error(_("Error creating modlog case"), exc_info=e)
+        except discord.errors.Forbidden:
+            return
+
+    @commands.Cog.listener()
+    async def on_member_remove(self, member: discord.Member):
+        guild = member.guild
+        if guild.id in self._server_mutes:
+            if member.id in self._server_mutes[guild.id]:
+                del self._server_mutes[guild.id][member.id]
+        for channel in guild.channels:
+            if channel.id in self._channel_mutes:
+                if member.id in self._channel_mutes[channel.id]:
+                    del self._channel_mutes[channel.id][member.id]
+        mute_role = await self.config.guild(guild).mute_role()
+        if not mute_role:
```

Comment:
This isn't a safe way to determine if someone was muted, data about mutes should be stored upon muting, it also only works if they were role muted instead of overwrites being applied.

Reply:
This was intentional when I started on this as I did not want to support permission based muting be applied when a user left and rejoined encouraging people use the role based model instead if they wanted to prevent people leaving and rejoining to avoid a mute.